File: csharp/Ice/invoke/PrinterI.cs
// Copyright (c) ZeroC, Inc.

using Demo;

internal class PrinterI : Ice.Blobject
{
    private class ReadValueCallback
    {
        public void Invoke(Ice.Value obj)
        {
            this.Obj = obj;
        }

        internal Ice.Value Obj;
    }

    public override bool ice_invoke(byte[] inParams, out byte[] outParams, Ice.Current current)
    {
        outParams = null;
        bool result = true;

        var communicator = current.adapter.getCommunicator();

        Ice.InputStream inStream = null;
        if (inParams.Length > 0)
        {
            inStream = new Ice.InputStream(communicator, inParams);
            inStream.startEncapsulation();
        }

        if (current.operation.Equals("printString"))
        {
            var message = inStream.readString();
            Console.WriteLine("Printing string `" + message + "'");
        }
        else if (current.operation.Equals("printStringSequence"))
        {
            var seq = StringSeqHelper.read(inStream);
            Console.Write("Printing string sequence {");
            for (int i = 0; i < seq.Length; ++i)
            {
                if (i > 0)
                {
                    Console.Write(", ");
                }
                Console.Write("'" + seq[i] + "'");
            }
            Console.WriteLine("}");
        }
        else if (current.operation.Equals("printDictionary"))
        {
            var dict = StringDictHelper.read(inStream);
            Console.Write("Printing dictionary {");
            bool first = true;
            foreach (var e in dict)
            {
                if (!first)
                {
                    Console.Write(", ");
                }
                first = false;
                Console.Write(e.Key + "=" + e.Value);
            }
            Console.WriteLine("}");
        }
        else if (current.operation.Equals("printEnum"))
        {
            var c = ColorHelper.read(inStream);
            Console.WriteLine("Printing enum " + c);
        }
        else if (current.operation.Equals("printStruct"))
        {
            var s = Structure.ice_read(inStream);
            Console.WriteLine("Printing struct: name=" + s.name + ", value=" + s.value);
        }
        else if (current.operation.Equals("printStructSequence"))
        {
            var seq = StructureSeqHelper.read(inStream);
            Console.Write("Printing struct sequence: {");
            for (int i = 0; i < seq.Length; ++i)
            {
                if (i > 0)
                {
                    Console.Write(", ");
                }
                Console.Write(seq[i].name + "=" + seq[i].value);
            }
            Console.WriteLine("}");
        }
        else if (current.operation.Equals("printClass"))
        {
            var cb = new ReadValueCallback();
            inStream.readValue(cb.Invoke);
            inStream.readPendingValues();
            var c = cb.Obj as C;
            Console.WriteLine("Printing class: s.name=" + c.s.name + ", s.value=" + c.s.value);
        }
        else if (current.operation.Equals("getValues"))
        {
            var c = new C(new Structure("green", Color.green));
            var outStream = new Ice.OutputStream(communicator);
            outStream.startEncapsulation();
            outStream.writeValue(c);
            outStream.writeString("hello");
            outStream.writePendingValues();
            outStream.endEncapsulation();
            outParams = outStream.finished();
        }
        else if (current.operation.Equals("throwPrintFailure"))
        {
            Console.WriteLine("Throwing PrintFailure");
            var ex = new PrintFailure("paper tray empty");
            var outStream = new Ice.OutputStream(communicator);
            outStream.startEncapsulation();
            outStream.writeException(ex);
            outStream.endEncapsulation();
            outParams = outStream.finished();
            result = false;
        }
        else if (current.operation.Equals("shutdown"))
        {
            current.adapter.getCommunicator().shutdown();
        }
        else
        {
            throw new Ice.OperationNotExistException(current.id, current.facet, current.operation);
        }

        // Make sure we read all in parameters
        inStream.endEncapsulation();
        return result;
    }
}

File: java/Ice/invoke/PrinterI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import com.zeroc.demos.Ice.invoke.Demo.*;

public class PrinterI implements com.zeroc.Ice.Blobject
{
    @Override
    public com.zeroc.Ice.Object.Ice_invokeResult ice_invoke(byte[] inParams, com.zeroc.Ice.Current current)
    {
        com.zeroc.Ice.Communicator communicator = current.adapter.getCommunicator();

        com.zeroc.Ice.InputStream in = new com.zeroc.Ice.InputStream(communicator, inParams);
        in.startEncapsulation();

        com.zeroc.Ice.Object.Ice_invokeResult r = new com.zeroc.Ice.Object.Ice_invokeResult();
        r.returnValue = true;

        if(current.operation.equals("printString"))
        {
            String message = in.readString();
            System.out.println("Printing string `" + message + "'");
        }
        else if(current.operation.equals("printStringSequence"))
        {
            String[] seq = in.readStringSeq();
            System.out.print("Printing string sequence {");
            for(int i = 0; i < seq.length; ++i)
            {
                if(i > 0)
                {
                    System.out.print(", ");
                }
                System.out.print("'" + seq[i] + "'");
            }
            System.out.println("}");
        }
        else if(current.operation.equals("printDictionary"))
        {
            java.util.Map<String, String> dict = StringDictHelper.read(in);
            System.out.print("Printing dictionary {");
            boolean first = true;
            for(java.util.Map.Entry<String, String> i : dict.entrySet())
            {
                if(!first)
                {
                    System.out.print(", ");
                }
                first = false;
                System.out.print(i.getKey() + "=" + i.getValue());
            }
            System.out.println("}");
        }
        else if(current.operation.equals("printEnum"))
        {
            Color c = Color.ice_read(in);
            System.out.println("Printing enum " + c);
        }
        else if(current.operation.equals("printStruct"))
        {
            Structure s = Structure.ice_read(in);
            System.out.println("Printing struct: name=" + s.name + ", value=" + s.value);
        }
        else if(current.operation.equals("printStructSequence"))
        {
            Structure[] seq = StructureSeqHelper.read(in);
            System.out.print("Printing struct sequence: {");
            for(int i = 0; i < seq.length; ++i)
            {
                if(i > 0)
                {
                    System.out.print(", ");
                }
                System.out.print(seq[i].name + "=" + seq[i].value);
            }
            System.out.println("}");
        }
        else if(current.operation.equals("printClass"))
        {
            class Holder
            {
                C obj;
            }
            final Holder h = new Holder();
            in.readValue(v -> h.obj = (C)v);
            in.readPendingValues();
            System.out.println("Printing class: s.name=" + h.obj.s.name + ", s.value=" + h.obj.s.value);
        }
        else if(current.operation.equals("getValues"))
        {
            C c = new C();
            c.s = new Structure();
            c.s.name = "green";
            c.s.value = Color.green;
            com.zeroc.Ice.OutputStream out = new com.zeroc.Ice.OutputStream(communicator);
            out.startEncapsulation();
            out.writeValue(c);
            out.writeString("hello");
            out.writePendingValues();
            out.endEncapsulation();
            r.outParams = out.finished();
        }
        else if(current.operation.equals("throwPrintFailure"))
        {
            System.out.println("Throwing PrintFailure");
            PrintFailure ex = new PrintFailure();
            ex.reason = "paper tray empty";
            com.zeroc.Ice.OutputStream out = new com.zeroc.Ice.OutputStream(communicator);
            out.startEncapsulation();
            out.writeException(ex);
            out.endEncapsulation();
            r.outParams = out.finished();
            r.returnValue = false;
        }
        else if(current.operation.equals("shutdown"))
        {
            current.adapter.getCommunicator().shutdown();
        }
        else
        {
            com.zeroc.Ice.OperationNotExistException ex = new com.zeroc.Ice.OperationNotExistException();
            ex.id = current.id;
            ex.facet = current.facet;
            ex.operation = current.operation;
            throw ex;
        }

        //
        // Verify we read all in parameters
        //
        in.endEncapsulation();
        return r;
    }
}
